ygen/enumgen.go

```diff
@@ -74,61 +74,186 @@ func newEnumGenState() *enumGenState {
-				entry: &yang.Entry{
-					Name: e.Name,
-					Type: &yang.YangType{
-						Name: e.Type.Name,
-						Kind: yang.Yenum,
-						Enum: t.Enum,
-					},
-					Annotation: map[string]interface{}{"valuePrefix": util.SchemaPathNoChoiceCase(e)},
-				},
-			}
+	definingModName := genutil.ParentModulePrettyName(e.Node)
+	// Since there can be many leaves that refer to the same typedef, then we do not generate
+	// a name for each of them, but rather use a common name, we use the non-CamelCase lookup
+	// as this is unique, whereas post-camelisation, we may have name clashes. Since a typedef
+	// does not have a 'path' in Goyang, so we synthesise one using the form
+	// module-name/typedef-name.
+	return fmt.Sprintf("%s/%s", definingModName, typeName), definingModName, typeName, nil
+}
+
+// enumLeafKey calculates a unique string key for the input enum leaf.  If
+// compressPaths is true, it also returns the compress name of the entry for
+// use in name generation, if needed.
+func (s *enumGenState) enumLeafKey(e *yang.Entry, compressPaths, noUnderscores, skipDedup bool) (string, string) {
+	// uniqueIdentifier is the unique identifier used to determine whether to
+	// define a new enum type for the input enum.
+	var uniqueIdentifier string
+	if skipDedup && !compressPaths {
+		// If not using compression and duplicating, then we use the
+		// entire path as the enum name, meaning every enum instance
+		// has its own definition. By using the entry's path as its
+		// identifier, we ensure this.
+		uniqueIdentifier = e.Path()
+	} else {
+		// In the other cases, de-duplication may happen, either
+		// through de-duping multiple usages or possibly through
+		// compression
+		uniqueIdentifier = enumIdentifier(e, compressPaths)
+	}
+
+	var compressName string
+	if compressPaths {
+		definingModName := genutil.ParentModulePrettyName(e.Node)
+		// If we compress paths then the name of this enum is of the form
+		// ModuleName_GrandParent_Leaf - we use GrandParent since Parent is
+		// State or Config so would not be unique. The proposed name is
+		// handed to genutil.MakeNameUnique to ensure that it does not clash with
+		// other defined names.
+		compressName = fmt.Sprintf("%s_%s_%s", yang.CamelCase(definingModName), yang.CamelCase(e.Parent.Parent.Name), yang.CamelCase(e.Name))
+		if noUnderscores {
+			compressName = strings.Replace(compressName, "_", "", -1)
 		}
 
-		es = append(es, en)
+		if skipDedup {
+			// If using compression and duplicating, then we add
+			// compressName to the uniqueIdentifier, meaning every
+			// enum instance in the compressed view of the schema
+			// has its own definition.  The base enum identity is
+			// still required to deal with collisions between
+			// compressed enum names when they describe different
+			// enums.
+			uniqueIdentifier += compressName
+		}
 	}
+	return uniqueIdentifier, compressName
+}
 
-	return es, nil
+// enumIdentifier takes in an enum entry and returns a unique identifier for
```

Comment:
Do we expect to call this in multiple places going forward? It seems like it's just called in a single place. Should we also be re-using the same function where there is an `enumeration` within a `union`?

Reply:
Indeed it's already being used by `enumeration` within a `union`, just that it's done indirectly via `enumName = s.resolveEnumName(e, compressPaths, noUnderscores, skipEnumDedup)` within `enumeratedUnionEntry()`.

I'll add a comment to `enumIdentifier` saying it can be used for union containing an enum as well.

Reply:
Thanks, SGTM.